Automattic/wp-calypso
# Editor Tracking - Track context of block events (template parts, reusable blocks, etc.)

#### Changes proposed in this Pull Request

Adds a `entity_context` property to block events when fired from within an entity controlling container.  This will help track how/what people are editing in the site editor.  This also extends to the post editor for reusable blocks.

Currently, the types of contexts in this PR are: "template", "core/template-part", "core/post-content", "core/query", and "core/block" (the reusable block).  When a block event happens, we use the `rootClientId` for the event and bubble up through the block hierarchy looking for one of these entity controller blocks listed above.  The first is returned as the context, and if none is found the default context is determined to be at the top level and is marked as "template" for the site editor and left `undefined` for the post editor as this would be a repeat of the already existing `post_type` property.

This should work for the following events:
* Move Up / Move Down
* Removing blocks
* Inserting blocks / patterns
* replacing blocks
* <s>drag and drop</s> - while I set this up to also work for the current drag-and-drop event tracking and allowed for both `from` and `to` type contexts, I found that drag-and-drop events are no longer tracking as expected in its current implementation. (opened an issue here https://github.com/Automattic/wp-calypso/issues/54353)

For Template Parts, if a variation is active it is appended to the end of the context string (ex. `core/template-part/header`) and a separate `template_part_id` property is also added to the event.

#### Testing instructions



* Load the site editor - test various block events in various contexts and verify their `entity_context` is listed as expected above.  For the top level, verify the context is "template".
* Load the post editor - add a reusable block or template part.  Verify block events within these containers populate the `entity_context` property as expected.  For block events the top level, verify there is no `entity_context` property.



Related to https://github.com/Automattic/wp-calypso/issues/53410
